File: solutions/python/bit-manipulation/maximum-product-of-word-lengths.py
# LeetCode 318 — Maximum Product of Word Lengths (Medium)
# Category: Bit Manipulation · Approach: Letter bitmask
# Time: O(n² + total) | Space: O(n)
# Source: https://leetcode.com/problems/maximum-product-of-word-lengths/

def max_product(words: list[str]) -> int:
    masks = []
    for w in words:
        m = 0
        for c in w:
            m |= 1 << (ord(c) - 97)
        masks.append(m)
    best = 0
    for i in range(len(words)):
        for j in range(i + 1, len(words)):
            if masks[i] & masks[j] == 0:
                best = max(best, len(words[i]) * len(words[j]))
    return best

File: solutions/typescript/bit-manipulation/maximum-product-of-word-lengths.ts
// LeetCode 318 — Maximum Product of Word Lengths (Medium)
// Category: Bit Manipulation · Approach: Letter bitmask
// Time: O(n² + total) | Space: O(n)
// Source: https://leetcode.com/problems/maximum-product-of-word-lengths/

function maxProduct(words: string[]): number {
  const masks = words.map((w) => {
    let m = 0;
    for (const c of w) m |= 1 << (c.charCodeAt(0) - 97);
    return m;
  });
  let best = 0;
  for (let i = 0; i < words.length; i++) {
    for (let j = i + 1; j < words.length; j++) {
      if ((masks[i] & masks[j]) === 0) {
        best = Math.max(best, words[i].length * words[j].length);
      }
    }
  }
  return best;
}
